React CFButton component

dgreisen-cfpb opened the pull request:
fully tested

contolini committed:
Merge branch 'slim-gruntfile' of https://github.com/Scotchester/cf-buttons into Scotchester-slim-gruntfile

contolini committed:
Merge branch 'slim-gruntfile' of https://github.com/Scotchester/cf-buttons into Scotchester-slim-gruntfile

contolini committed:
Merge branch 'gh-pages' of github.com:cfpb/cf-buttons into gh-pages

contolini committed:
Merge branch 'gh-pages' of github.com:cfpb/cf-buttons into gh-pages

contolini committed:
Merge branch 'gh-pages' of github.com:cfpb/cf-buttons into gh-pages

dgreisen-cfpb committed:
attempt to get tests working

contolini committed:
Merge branch 'disaster' of github.com:dgreisen-cfpb/cf-buttons into disaster

contolini committed:
:expressionless:

dgreisen-cfpb committed:
reimplemented with gulp

dgreisen-cfpb committed:
completed tests for button; changed name

Scotchester commented:
Gonna need some more info on this one :)

Scotchester commented:
@dgreisen-cfpb  and I spoke by phone and agreed that we should try to get the coverage functionality working in Grunt or via @contolini's tool-less method before accepting this PR.

dgreisen-cfpb committed:
removed gulp

dgreisen-cfpb committed:
cleanup; util fns; standardize with icon;

decided that 'config' makes more sense than 'type' for component
configuration. created util function to create className from
config, className and default classes. Will probably want to pull
this (and omit) out into a helper repo.

updated tests appropriately.

dgreisen-cfpb committed:
added docs

dgreisen-cfpb committed:
re-added accidentally removed grunt

dgreisen-cfpb committed:
removed unused test

imuchnik commented:
:+1:  I like it.

ascott1 commented:
This looks great @dgreisen-cfpb. :+1: from me.

Scotchester commented:
:+1: The JS needs some linting, but I don't blame you for that, since the linter isn't in the code yet.

dgreisen-cfpb commented:
@Scotchester clarified offline that linter is not blocker. So since it has unanimous approval, @Scotchester: want to do the honors and merge?

Scotchester commented:
:shipit:

Scotchester commented:
WAIT! Sorry. One more thing. We need to update the version number in `bower.json` and update the `CHANGELOG`.

dgreisen-cfpb committed:
bower ver-bump; changelog

dgreisen-cfpb commented:
@Scotchester: done

Scotchester commented:
:+1: Thanks!

Scotchester commented:
@dgreisen-cfpb Can you please tag and release? Let me know if you have questions about that.

cfarm commented:
@dgreisen-cfpb ditto @Scotchester! we're using this for OAH :)

ascott1 commented:
Nice @cfarm! Also check out https://github.com/ascott1/cf-icons/blob/react/src/CFIcon.jsx, which @dgreisen-cfpb and I were working on last week.

anselmbradford commented on src/CFButton.jsx:
```diff
@@ -0,0 +1,102 @@
+    };
+  },
+  render: function() {
+    var rightIcon, leftIcon
```
Missing ending semicolon;

anselmbradford commented on src/CFButton.jsx:
```diff
@@ -0,0 +1,102 @@
+  render: function() {
+    var rightIcon, leftIcon
+
+    var btnClassName = genClassName('btn', 'btn', this.props)
```
Missing ending semicolon;

anselmbradford commented on src/CFButton.jsx:
```diff
@@ -0,0 +1,102 @@
+  'warning',
+];
+
+function genClassName(requiredClasses, configPrefix, props) {
```
Consider underscore prefixing "private" methods localized to only this module: `_genClassName`.

dgreisen-cfpb replied:
this will (hopefully) soon be pulled out into a helper repo.

anselmbradford commented on src/CFButton.jsx:
```diff
@@ -0,0 +1,102 @@
+}
+
+var CFButton = React.createClass({
+  /*
```
Use jsdoc comments http://usejsdoc.org/

anselmbradford commented on src/CFButton.jsx:
```diff
@@ -0,0 +1,102 @@
+
+    if (this.props.href) {
+      return <a className={btnClassName} {...buttonProps}>{leftIcon}{this.props.children}{rightIcon}</a>
+    } else {
```
else statement is unneeded.

anselmbradford commented on src/CFButton.jsx:
```diff
@@ -0,0 +1,102 @@
+});
+
+function omit(obj) {
+  var omitted = Array.prototype.slice.call(arguments, 1);
```
See warning on https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments

anselmbradford replied:
May not matter due to the length being sliced.

anselmbradford commented on src/CFButton.jsx:
```diff
@@ -0,0 +1,102 @@
+      out[prop] = obj[prop];
+    }
+  }
+  return out
```
Missing semicolon;

anselmbradford commented on src/CFButton.jsx:
```diff
@@ -0,0 +1,102 @@
+  var omitted = Array.prototype.slice.call(arguments, 1);
+  var out = {};
+  var props = Object.getOwnPropertyNames(obj);
+  for (var i=0;i<props.length;i++) {
```
Length lookup each loop can be prevented with:

``` js
for ( var i = 0, len = props.length; i < len; i++ ) {
```

dgreisen-cfpb commented:
@Scotchester: I do not know how to tag and release. Probably not the best person to do that, either.
